### PR TITLE
Add tunnel state machine

### PR DESCRIPTION
**Pull request by pronebird**
This is a draft of PR that attempts to:

- Put together a tunnel management using a state machine
- Attempt to separate dependencies from mix/wg tunnel setup

In principle the tunnel state machine is comprised of individual objects each implementing `TunnelStateHandler` for polling and handling events. There are 5 different states: disconnected, connecting, connected, disconnecting and error state. Each is responsible for different aspects of tunnel lifecycle. In response to event, the tunnel state handler may either remain in the same state or instruct the state machine to move to the next state.

Commands can be passed to state machine over the channel. These commands can be handled differently by each state handler. 

As the tunnel state machine transitions through different states, the higher level tunnel state is passed back via the channel to the outside observer in order to keep track of what's going on.

- - -
This change is [Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/1109)

### CODE REVIEW

File: nym-vpn-core/crates/nym-vpn-lib/src/tunnel_state_machine/mixnet_route_handler.rs [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    }
+
+    pub async fn stop(mut self) {
+        #[cfg(windows)]
+        self.route_manager.stop();
+
+        #[cfg(not(windows))]
+        self.route_manager.stop().await;
+
+        tokio::task::spawn_blocking(|| drop(self.route_manager));
```
Comment (octol) [inline nym-vpn-core/crates/nym-vpn-lib/src/tunnel_state_machine/mixnet_route_handler.rs:84]: is it worth joining this task before returning from this function?

Comment (pronebird) [reply nym-vpn-core/crates/nym-vpn-lib/src/tunnel_state_machine/mixnet_route_handler.rs:84]: Good catch. Of course we should destroy route manager before returning.

**Review by doums — COMMENTED** [review]
_out of scope_
Remember me last time I had to implement a finite state machine for one of my side project (https://github.com/doums/bato/blob/master/src/fsm.rs if you're interested)

File: nym-vpn-core/crates/nym-vpn-lib/src/mobile/runner.rs [hunk 2]
```diff
@@ -185,6 +183,9 @@ impl WgTunnelRunner {
             }
         }
 
+        // Shutdown mixnet client.
+        mixnet_client.disconnect().await;
```
Comment (octol) [inline nym-vpn-core/crates/nym-vpn-lib/src/mobile/runner.rs:187]: Note that this is a no-op if it has been built with and external task manager

Comment (pronebird) [reply nym-vpn-core/crates/nym-vpn-lib/src/mobile/runner.rs:187]: So shutting down the task manager should be enough then right?

Comment (octol) [reply nym-vpn-core/crates/nym-vpn-lib/src/mobile/runner.rs:187]: yes, signal disconnect and possible also wait for it to finish disconnecting

**Review by octol — COMMENTED** [review]
> Reviewed 1 of 12 files at r1, 2 of 21 files at r2, 2 of 26 files at r4, 17 of 29 files at r6, 24 of 26 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/1109)* status: 46 of 57 files reviewed, 4 unresolved discussions (waiting on @pronebird)

___
*[`nym-vpn-core/crates/nym-vpn-lib/src/tunnel_state_machine/dns_handler.rs` line 81 at r7](https://reviewable.io/reviews/nymtech/nym-vpn-client/1109#-O8a8mNkAPVWqU7q7zOV:-O8a8mNkAPVWqU7q7zOW:b-qioynh) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/652f74fa4fbe72beeb7f3fd66c22daf92db6a817/nym-vpn-core/crates/nym-vpn-lib/src/tunnel_state_machine/dns_handler.rs#L81)):*
> ```XML
> 
> #[derive(Debug)]
> pub struct Error {
> ```

Tell me more about this Error, why not just send the source error? Doesn't look like we're adding any context?
___
*[`nym-vpn-core/crates/nym-vpn-lib/src/tunnel_state_machine/firewall_handler.rs` line 38 at r7](https://reviewable.io/reviews/nymtech/nym-vpn-client/1109#-O8a93Xo3DpXmcgc003c:-O8a93Xo3DpXmcgc003d:b-y1wn9k) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/652f74fa4fbe72beeb7f3fd66c22daf92db6a817/nym-vpn-core/crates/nym-vpn-lib/src/tunnel_state_machine/firewall_handler.rs#L38)):*
> ```XML
> 
> #[derive(Debug)]
> pub struct Error {
> ```

Same here as for the dns error
___
*[`nym-vpn-core/crates/nym-vpn-lib/src/tunnel_state_machine/mod.rs` line 78 at r7](https://reviewable.io/reviews/nymtech/nym-vpn-client/1109#-O8a9Jqh1dRsxZDcKPto:-O8a9Jqh1dRsxZDcKPtp:bhrc637) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/652f74fa4fbe72beeb7f3fd66c22daf92db6a817/nym-vpn-core/crates/nym-vpn-lib/src/tunnel_state_machine/mod.rs#L78)):*
> ```XML
> 
>     /// Failure to establish mixnet connection.
>     EstablishMixnetConnection,
> ```

So this was something that was discussed earlier. I think we can merge this PR without fleshing out the error handling, but I think it would be a mistake to not propagate the error source and context and I hope that's something we can do before switching over the daemon to use the state machine.
"Developer me" will look at the logs, but "User me" would never do that.
___
*[`nym-vpn-core/crates/nym-vpn-lib/src/tunnel_state_machine/mod.rs` line 242 at r7](https://reviewable.io/reviews/nymtech/nym-vpn-client/1109#-O8aAQhNAK0pZeAWyMtb:-O8aAQhNAK0pZeAWyMtc:b-xmhesl) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/652f74fa4fbe72beeb7f3fd66c22daf92db6a817/nym-vpn-core/crates/nym-vpn-lib/src/tunnel_state_machine/mod.rs#L242)):*
> ```XML
>             }
>             Self::ConnectMixnetTunnel(_) | Self::ConnectMixnetClient(_) => {
>                 // todo: add detail
> ```

Yep. We can do that in a later PR

**Review by octol — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/1109)* status: 46 of 57 files reviewed, 5 unresolved discussions (waiting on @pronebird)

___
*[`rustfmt.toml` line 1 at r7](https://reviewable.io/reviews/nymtech/nym-vpn-client/1109#-O8a3poXBedLnQrjejak:-O8a3poXBedLnQrjejal:b8fzwlq) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/652f74fa4fbe72beeb7f3fd66c22daf92db6a817/rustfmt.toml#L1)):*
> ```TOML
> imports_granularity = "Crate"
> ```

So I'm still not sure how I feel about this, and we probably should involve more people before the decision to start using nightly toolchain is taken

**Review by octol — COMMENTED** [review]
> Reviewed 1 of 21 files at r2, 1 of 6 files at r3, 1 of 26 files at r4, 6 of 29 files at r6, 2 of 26 files at r7.
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/1109)* status: all files reviewed, 4 unresolved discussions (waiting on @pronebird)

**Review by pronebird — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/1109)* status: all files reviewed, 4 unresolved discussions (waiting on @octol)

___
*[`rustfmt.toml` line 1 at r7](https://reviewable.io/reviews/nymtech/nym-vpn-client/1109#-O8a3poXBedLnQrjejak:-O8aNocc2JckFOhIih2K:bc7fafi) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/652f74fa4fbe72beeb7f3fd66c22daf92db6a817/rustfmt.toml#L1)):*
<details><summary><i>Previously, octol (Jon Häggblad) wrote…</i></summary><blockquote>

So I'm still not sure how I feel about this, and we probably should involve more people before the decision to start using nightly toolchain is taken
</blockquote></details>

Technically apart from warnings it has zero impact or unless you attempt to format using nightly. The rule itself produces a more compact import statements collapsed at crate level, which looks cleaner. 

AFAIK you've already formatted part of the code with this rule, but since there is no recommendation nor config to produce such formatting, then we're still at the crossroads and people do things their own way which is sub-optimal. We should certainly raise a discussion and come to agreement.
___
*[`nym-vpn-core/crates/nym-vpn-lib/src/mobile/runner.rs` line 187 at r2](https://reviewable.io/reviews/nymtech/nym-vpn-client/1109#gh-1759416860:-O8aPrdI9h-oshh-okvH:b-u6tn2j) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/8421f26e7c884ec6b7085d23ea86c0a93b38adc1/nym-vpn-core/crates/nym-vpn-lib/src/mobile/runner.rs#L187)):*
<details><summary><i>Previously, octol (Jon Häggblad) wrote…</i></summary><blockquote>

yes, signal disconnect and possible also wait for it to finish disconnecting
</blockquote></details>

Let me delete this
___
*[`nym-vpn-core/crates/nym-vpn-lib/src/tunnel_state_machine/dns_handler.rs` line 81 at r7](https://reviewable.io/reviews/nymtech/nym-vpn-client/1109#-O8a8mNkAPVWqU7q7zOV:-O8aL2aD4RO0qldjj1nb:b-z3xw7w) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/652f74fa4fbe72beeb7f3fd66c22daf92db6a817/nym-vpn-core/crates/nym-vpn-lib/src/tunnel_state_machine/dns_handler.rs#L81)):*
<details><summary><i>Previously, octol (Jon Häggblad) wrote…</i></summary><blockquote>

Tell me more about this Error, why not just send the source error? Doesn't look like we're adding any context?
</blockquote></details>

It doesn't add any context but shields the rest of state machine from talpid errors
___
*[`nym-vpn-core/crates/nym-vpn-lib/src/tunnel_state_machine/firewall_handler.rs` line 38 at r7](https://reviewable.io/reviews/nymtech/nym-vpn-client/1109#-O8a93Xo3DpXmcgc003c:-O8aLh9P8qj9CJmTeMjw:bkdn709) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/652f74fa4fbe72beeb7f3fd66c22daf92db6a817/nym-vpn-core/crates/nym-vpn-lib/src/tunnel_state_machine/firewall_handler.rs#L38)):*
<details><summary><i>Previously, octol (Jon Häggblad) wrote…</i></summary><blockquote>

Same here as for the dns error
</blockquote></details>

Very much just hiding impl. details.
___
*[`nym-vpn-core/crates/nym-vpn-lib/src/tunnel_state_machine/mod.rs` line 78 at r7](https://reviewable.io/reviews/nymtech/nym-vpn-client/1109#-O8a9Jqh1dRsxZDcKPto:-O8aLmODCEE8GCgzpopR:bmkyvtq) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/652f74fa4fbe72beeb7f3fd66c22daf92db6a817/nym-vpn-core/crates/nym-vpn-lib/src/tunnel_state_machine/mod.rs#L78)):*
<details><summary><i>Previously, octol (Jon Häggblad) wrote…</i></summary><blockquote>

So this was something that was discussed earlier. I think we can merge this PR without fleshing out the error handling, but I think it would be a mistake to not propagate the error source and context and I hope that's something we can do before switching over the daemon to use the state machine.
"Developer me" will look at the logs, but "User me" would never do that.
</blockquote></details>

This is kept intentionally flat and simple for the following reasons:

- The work to inspect the error chain and conclude what went wrong is the responsibility of this module as it has much deeper knowledge of the implementation detail.
- The consumer of this module can then match the error reason and produce the localized messages to display to user. 
- It can be copied as the type is primitive.
- Note that once we add reconnection loop, errors related to connectivity are likely going away as we'd keep picking different gateways upon failure to connect.

**Review by pronebird — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/1109)* status: 53 of 57 files reviewed, 4 unresolved discussions (waiting on @octol)

___
*[`rustfmt.toml` line 1 at r7](https://reviewable.io/reviews/nymtech/nym-vpn-client/1109#-O8a3poXBedLnQrjejak:-O8amqTm-FHqw7Tgl1aV:b-886gsx) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/652f74fa4fbe72beeb7f3fd66c22daf92db6a817/rustfmt.toml#L1)):*
<details><summary><i>Previously, pronebird (Andrej Mihajlov) wrote…</i></summary><blockquote>

Technically apart from warnings it has zero impact or unless you attempt to format using nightly. The rule itself produces a more compact import statements collapsed at crate level, which looks cleaner. 

AFAIK you've already formatted part of the code with this rule, but since there is no recommendation nor config to produce such formatting, then we're still at the crossroads and people do things their own way which is sub-optimal. We should certainly raise a discussion and come to agreement.
</blockquote></details>

Fixed

**Review by octol — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/1109)* status: 53 of 57 files reviewed, all discussions resolved (waiting on @pronebird)

___
*[`nym-vpn-core/crates/nym-vpn-lib/src/mobile/runner.rs` line 187 at r2](https://reviewable.io/reviews/nymtech/nym-vpn-client/1109#gh-1759416860:-O8awRbB4xpLWYU8hG5U:bb7f2mf) ([raw file](https://github.com/nymtech/nym-vpn-client/blob/8421f26e7c884ec6b7085d23ea86c0a93b38adc1/nym-vpn-core/crates/nym-vpn-lib/src/mobile/runner.rs#L187)):*
<details><summary><i>Previously, pronebird (Andrej Mihajlov) wrote…</i></summary><blockquote>

Let me delete this
</blockquote></details>

done

**Review by octol — APPROVED** [review]
> Reviewed 4 of 4 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/nymtech/nym-vpn-client/1109)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @pronebird)

**Pull request merged**